unittests/Interpreter/InterpreterTests.cpp

```diff
@@ -25,345 +27,69 @@
-    if (!a.hasValue())
-      return b;
-    if (!b.hasValue())
-      return a;
-
-    // FIXME: Handle other ValueKind types. how?
-    assert(a.hasValue() && b.hasValue());
-    auto aVal = a.getValueKB();
-    auto bVal = b.getValueKB();
-    assert(aVal.getBitWidth() == bVal.getBitWidth() && "Can only merge KnownBits of same width.");
-
-    KnownBits res(aVal.getBitWidth());
-    for (unsigned i = 0; i < aVal.getBitWidth(); i++) {
-      if (aVal.One[i] == bVal.One[i] && bVal.One[i])
-        res.One.setBit(i);
-      if (aVal.Zero[i] == bVal.Zero[i] && aVal.Zero[i])
-        res.Zero.setBit(i);
-    }
+TEST(InterpreterTests, KBTransferFunctions) {
+  KBTesting kbObj;
+  ASSERT_TRUE(kbObj.testFn(Inst::Add));
+  ASSERT_TRUE(kbObj.testFn(Inst::AddNSW));
+  ASSERT_TRUE(kbObj.testFn(Inst::Sub));
+  ASSERT_TRUE(kbObj.testFn(Inst::SubNSW));
+  ASSERT_TRUE(kbObj.testFn(Inst::Mul));
+  ASSERT_TRUE(kbObj.testFn(Inst::UDiv));
+  ASSERT_TRUE(kbObj.testFn(Inst::URem));
+  ASSERT_TRUE(kbObj.testFn(Inst::And));
+  ASSERT_TRUE(kbObj.testFn(Inst::Or));
+  ASSERT_TRUE(kbObj.testFn(Inst::Xor));
+  ASSERT_TRUE(kbObj.testFn(Inst::Shl));
+  ASSERT_TRUE(kbObj.testFn(Inst::LShr));
+  ASSERT_TRUE(kbObj.testFn(Inst::AShr));
+  ASSERT_TRUE(kbObj.testFn(Inst::Eq));
+  ASSERT_TRUE(kbObj.testFn(Inst::Ne));
+  ASSERT_TRUE(kbObj.testFn(Inst::Ult));
+  ASSERT_TRUE(kbObj.testFn(Inst::Slt));
+  ASSERT_TRUE(kbObj.testFn(Inst::Ule));
+  ASSERT_TRUE(kbObj.testFn(Inst::Sle));
+}
 
-    return EvalValueKB(res);
-  }
+TEST(InterpreterTests, CRTransferFunctions) {
+  CRTesting crObj;
+  ASSERT_TRUE(crObj.testFn(Inst::And));
+  ASSERT_TRUE(crObj.testFn(Inst::Or));
+}
 
-  KnownBits setLowest(KnownBits x) {
-    for (int i = 0; i < x.getBitWidth(); i++) {
-      if (!x.Zero[i] && !x.One[i]) {
-        x.One.setBit(i);
-        return x;
+TEST(InterpreterTests, KBCRReduction) {
+  ConstantRange CR(WIDTH, /*isFullSet=*/false);
+  KnownBits KB(WIDTH);
+  do {
+    do {
+      KnownBits CalculatedKB = KB;
+      ConstantRange CalculatedCR = CR;
+      improveKBCR(CalculatedKB, CalculatedCR);
+
+      KnownBits ExhaustiveKB = KB;
+      ConstantRange ExhaustiveCR = CR;
+      TestingUtil::exhaustiveKBCRReduction(ExhaustiveKB, ExhaustiveCR);
+
+      if (KnownBitsAnalysis::isConflictingKB(CalculatedKB, ExhaustiveKB)) {
+	outs() << "Unsound!! CR KB reduction for KB\n";
+	outs() << "Original KB: " << KnownBitsAnalysis::knownBitsString(KB) << "\n";
+	outs() << "Original CR: " << CR << "\n";
+	outs() << "CalculatedKB: " << KnownBitsAnalysis::knownBitsString(CalculatedKB) << '\n';
+	outs() << "ExhaustiveKB: " << KnownBitsAnalysis::knownBitsString(ExhaustiveKB) << '\n';
+	ASSERT_TRUE(false);
       }
-    }
-    report_fatal_error("faulty setLowest!");
-  }
 
-  KnownBits clearLowest(KnownBits x) {
-    for (int i = 0; i < x.getBitWidth(); i++) {
-      if (!x.Zero[i] && !x.One[i]) {
-        x.Zero.setBit(i);
-        return x;
+      if (!ExhaustiveCR.contains(CalculatedCR)) {
```

Comment:
are you sure this is right? shouldn't it be the CalculatedCR that contains the ExhaustiveCR?

the way to think about this is: the ExhaustiveCR must be the smallest one, so why would it contain a worse one?

Reply:
also this contains property (whichever way it should go) doesn't necessarily hold since the constantrange abstract domain isn't a lattice

Reply:
oh, yes, you are right. Let me fix that.